### PR TITLE
Add sysctlInit to OpenSearch Chart

### LINKED ISSUE #87: forbidden sysctl: "vm.max_map_count" not whitelisted
Hi,

i've tried to install opensearch via helm, but pods are pending with this message:
forbidden sysctl: "vm.max_map_count" not whitelisted

I've tried to override the security settings, but did not help:
```console
helm upgrade --install opensearch opensearch/opensearch --set sysctl.enabled=true --set podSecurityContext.runAsUser=0 --set securityContext.runAsNonRoot=false >test.yml
```

Im try to migrate from ES, but theirs helm using a bit different method:
they are using a fully privileges initcontainer to set the vm.max_map_count:
```
   initContainers:
154       - name: configure-sysctl
155         securityContext:
156           runAsUser: 0
157           privileged: true
158         image: "docker.elastic.co/elasticsearch/elasticsearch:7.14.0"
159         imagePullPolicy: "IfNotPresent"
160         command: ["sysctl", "-w", "vm.max_map_count=262144"]
161         resources:
```

What would be the best practice? Others does not have this situation?

thanks

### PR DESCRIPTION
**Pull request by amirgo1**
Signed-off-by: amirgo1 <amirgoro1@gmail.com>

### Description
Add option to enable the use of `sysctlInitContainer` to set sysctl vm.max_map_count through privileged `initContainer`.
 
### Issues Resolved
#87 
 
### Check List
- [x] Commits are signed per the DCO using --signoff

For any changes to files within Helm chart directories:
- [x] Helm chart version bumped
- [x] Helm chart `CHANGELOG.md` updated to reflect change

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
Hi @amirgo1 thanks for contribution,
please resolve the conflicts so we can try to start the workflow and review.
Thanks.

**Comment by DandyDeveloper** [comment]
@peterzhuamazon @amirgo1 Before we approve this as well, we really want to make sure there's a consensus from everyone that this is how we want to do this. 

It's a little worrying that we need to do this via privileged containers that are changing the host OS sysctl settings. 


@smlx @mprimeaux I'll just invite everyone to get a "Yes, lets do this". Maybe thumbs up / Thumbs down?

**Comment by peterzhuamazon** [comment]
Adding @prudhvigodithi to the discussion please see above msg.
Thanks.

File: charts/opensearch/templates/statefulset.yaml
Comment (smlx) [inline charts/opensearch/templates/statefulset.yaml:235]: For consistency could we change this to work the same way as the persistence container?
- `sysctlInit.enabled` to enable the functionality
- `sysctlInit.image` and `sysctlInit.imageTag` to control those field.
- default to `busybox:latest` since we don't need any opensearch scripts.

File: charts/opensearch/templates/statefulset.yaml
Comment (smlx) [inline charts/opensearch/templates/statefulset.yaml:239]: I don't think we should write this every time the init container runs unless it is necessary. Maybe some logic like this to check the value and only write it if required? After all, you don't want to _reduce_ this value if some other service needs it to be even higher.
```
  command:
  - sh
  - -c
  - |
    set -xe
    DESIRED="{{ .Values.sysctl.VmMaxMapCount }}"
    CURRENT=$(sysctl -n vm.max_map_count)
    if [ "$DESIRED" -gt "$CURRENT" ]; then
      sysctl -w vm.max_map_count=$DESIRED
    fi
```

File: charts/opensearch/README.md
Comment (smlx) [inline charts/opensearch/README.md:99]: I don't think "requires privilege" is enough detail here because the situation is quite complex. Could you add a bit more detail about how this won't work unless you add an argument to the kubelet, and maybe link through to https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/?

File: charts/opensearch/Chart.yaml
Comment (smlx) [inline charts/opensearch/Chart.yaml:18]: Because the API (values file) is changing this is at least a minor version bump, rather than a bugfix.

**Review by smlx — CHANGES_REQUESTED** [review]
I'm ok with the concept for this since it's the only practical solution in many k8s environments.

File: charts/opensearch/README.md
Comment (prudhvigodithi) [reply charts/opensearch/README.md:99]: We only need `privileged: true` for `sysctlInitContainer` (that is being pushed via this PR) and not for general `sysctl` and this gets added via  `corev1.PodSecurityContext`, so should we mention this line `This requires privilege` ?
@smlx @amirgo1

**Comment by peterzhuamazon** [comment]
Hi @amirgo1 please resolve the conflicts as well as checking out @smlx review so we can proceed.
Thanks.

File: charts/opensearch/README.md
Comment (amirgo1) [reply charts/opensearch/README.md:99]: @prudhvigodithi I agree with you and think `.Values.sysctl` need to be removed along with `PodSecurityPolicy` as it shouldn't work at all.
vm.max_map_count sysctl is not namespaced, so it cannot be configured through securityContext.

As explained here:
https://github.com/opensearch-project/helm-charts/issues/87#issuecomment-942738258

@smlx - what do you think?

File: charts/opensearch/Chart.yaml
Comment (TheAlgo) [reply charts/opensearch/Chart.yaml:18]: Yes this should be a minor version bump. I agree with @smlx

**Review by TheAlgo — COMMENTED** [review]
I am also okay with this option since I see this as one of the viable options for configuring this property. We have seen some issues with this option when the charts were in the initial phase as well.

File: charts/opensearch/templates/statefulset.yaml
Comment (TheAlgo) [reply charts/opensearch/templates/statefulset.yaml:239]: Yup this will be un-necessary to do it. To write if only required is a good strategy

File: charts/opensearch/values.yaml
Comment (TheAlgo) [inline charts/opensearch/values.yaml:413]: Good to add these options in the CI as well so that it is tested everytime : https://github.com/opensearch-project/helm-charts/tree/main/charts/opensearch/ci

File: charts/opensearch/README.md
Comment (smlx) [reply charts/opensearch/README.md:99]: I think that removing `.Values.sysctl` is unrelated to the changes in this PR. It belongs in another PR, and it's a breaking change so would need a major version bump.

**Comment by peterzhuamazon** [comment]
@TheAlgo @smlx @prudhvigodithi please approve this if you are ok with the change, I will resolve the conflicts later.
Thanks.

**Comment by peterzhuamazon** [comment]
@amirgo1 could you check @TheAlgo comments on this?
Thanks.

File: charts/opensearch/templates/statefulset.yaml
Comment (amirgo1) [reply charts/opensearch/templates/statefulset.yaml]: Updated.

File: charts/opensearch/README.md
Comment (amirgo1) [reply charts/opensearch/README.md]: I took this out of the documentation (it wasn't there before).

File: charts/opensearch/values.yaml
Comment (amirgo1) [reply charts/opensearch/values.yaml:420]: I updated all the values files :)

**Comment by amirgo1** [comment]
@peterzhuamazon @TheAlgo my apologies for the delay. I updated the PR

**Comment by amirgo1** [comment]
@TheAlgo, could you please review the changes I made?

File: charts/opensearch/ci/ci-values.yaml
Comment (TheAlgo) [inline charts/opensearch/ci/ci-values.yaml:405]: Can we enable it in one of the CI(s)?

File: charts/opensearch/CHANGELOG.md
Comment (TheAlgo) [inline charts/opensearch/CHANGELOG.md:16]: This needs to become `2.4.2` as there is a PR #279 which is taking up the previous version

File: charts/opensearch/CHANGELOG.md
Comment (TheAlgo) [inline charts/opensearch/CHANGELOG.md:18]: Please update the last section of the CHANGELOG which contains the macro for the sections of Unreleased and version specifics

**Review by TheAlgo — CHANGES_REQUESTED** [review]

**Comment by amirgo1** [comment]
@TheAlgo I've made the changes

File: charts/opensearch/templates/statefulset.yaml
Comment (TheAlgo) [inline charts/opensearch/templates/statefulset.yaml:248]: Can we add sample placeholders in `values.yaml` for `vmMaxMapCount`

**Review by TheAlgo — CHANGES_REQUESTED** [review]

**Comment by TheAlgo** [comment]
@smlx @prudhvigodithi Can you guys please review it when free? Thanks

**Comment by TheAlgo** [comment]
@amirgo1 Any update on this PR?

**Comment by amirgo1** [comment]
@TheAlgo Fixed comment and bumped version

**Review by TheAlgo — APPROVED** [review]
LGTM!

**Comment by amirgo1** [comment]
Resolved changelog conflict @prudhvigodithi @TheAlgo

**Comment by TheAlgo** [comment]
@prudhvigodithi @DandyDeveloper Please review when free. Thanks

**Comment by TheAlgo** [comment]
@amirgo1 Can you check the lint errors?

**Comment by peterzhuamazon** [comment]
Seems like no update from @amirgo1 I will fix the link for him.

**Comment by peterzhuamazon** [comment]
Thanks @amirgo1 for the contribution and patience, we will merge this PR now.

**Review by peterzhuamazon — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
Backport 1.x: https://github.com/opensearch-project/helm-charts/pull/349

**Pull request merged**
